stats/opencensus/opencensus.go

```diff
@@ -78,15 +83,89 @@ func ServerOption(to TraceOptions) grpc.ServerOption {
 }
 
 // unaryInterceptor handles per RPC context management. It also handles per RPC
-// tracing and stats.
+// tracing and stats, and records the latency for the full RPC call.
 func unaryInterceptor(ctx context.Context, method string, req, reply interface{}, cc *grpc.ClientConn, invoker grpc.UnaryInvoker, opts ...grpc.CallOption) error {
-	return invoker(ctx, method, req, reply, cc, opts...)
+	startTime := time.Now()
+	err := invoker(ctx, method, req, reply, cc, opts...)
+	callLatency := float64(time.Since(startTime)) / float64(time.Millisecond)
+
+	var st string
+	if err != nil {
+		s := status.Convert(err)
+		st = canonicalString(s.Code())
+	} else {
+		st = "OK"
+	}
+
+	ocstats.RecordWithOptions(ctx,
+		ocstats.WithTags(
+			tag.Upsert(keyClientMethod, removeLeadingSlash(method)),
+			tag.Upsert(keyClientStatus, st),
+		),
+		ocstats.WithMeasurements(
+			clientAPILatency.M(callLatency),
+		),
+	)
+
+	return err
+}
+
+// wrappedStream wraps a grpc.ClientStream to intercept the RecvMsg call to take
+// latency measurements on RPC Completion.
+type wrappedStream struct {
+	grpc.ClientStream
+	// The timestamp of when this stream started.
+	startTime time.Time
+	// the method for this stream.
+	method string
+}
+
+func newWrappedStream(s grpc.ClientStream) grpc.ClientStream {
+	return &wrappedStream{ClientStream: s}
+}
+
+func (ws *wrappedStream) RecvMsg(m interface{}) error {
```

Comment:
Streams can end in other ways than this.

https://github.com/grpc/grpc-go/blob/dba26e15a07f43875ccf806a2dd6cbcbc1c12eab/stream.go#L141-L150

In 1/2/4, this will never happen, and our metric will not be recorded.

We need a strategy for these.  (2) requires a goroutine to monitor the context which is inconvenient at the very least.  But I'm thinking the `stats.End` event can potentially be leveraged here.

Reply:
Discussed offline; all 4 of these cases call into ClientStream finish(), so switching to a callInfo callback called there handles all of these cases.